techniques/applications/repoGpgKeyManagement/1.0/repoGpgKeyManagement.st

```diff
@@ -0,0 +1,135 @@
+      delete     => tidy,
+      ifvarclass => "repo_gpg_key_${index_repo_gpg_keys}_imported";
+
+  commands:
+
+   debian|ubuntu::
+
+     "/usr/bin/apt-key fingerprint | sed -rn 's/^pub\s+.{5}\/([A-F0-9]{8}) .+/\+repo_gpg_key_\1_present/gp;' | tr '[:upper:]' '[:lower:]'"
+       comment => "We use this module instead of looking up each gpg key, because each call takes about 3-5 seconds...",
+       contain => outputable,
+       module  => "true";
+
+    "/usr/bin/apt-key add ${g.rudder_var_tmp}/gpg_key_${index_repo_gpg_keys}_content"
+      ifvarclass => "repo_gpg_key_${index_repo_gpg_keys}_action_add.!repo_gpg_key_${repo_gpg_hash_lc[${index_repo_gpg_keys}]}_present.gpg_key_file_${index_repo_gpg_keys}_created",
+      classes    => if_else("repo_gpg_key_${index_repo_gpg_keys}_imported", "repo_gpg_key_${index_repo_gpg_keys}_import_error");
+
+    "/usr/bin/apt-key del ${repo_gpg_hash_uc[${index_repo_gpg_keys}]}"
+      comment    => "This removes the GPG key from APT if it is present and needs to be removed",
+      ifvarclass => "repo_gpg_key_${index_repo_gpg_keys}_action_remove.repo_gpg_key_${repo_gpg_hash_lc[${index_repo_gpg_keys}]}_present",
+      classes    => if_else("repo_gpg_key_${index_repo_gpg_keys}_removed", "repo_gpg_key_${index_repo_gpg_keys}_remove_error");
+
+   redhat|SuSE::
+    "/bin/rpm --quiet -q gpg-pubkey-${repo_gpg_hash_lc[${index_repo_gpg_keys}]}"
+      comment    => "This checks if each GPG key is installed or is missing, and defines appropriate classes for this to work with",
+      classes    => if_ok("repo_gpg_key_${repo_gpg_hash_lc[${index_repo_gpg_keys}]}_present");
+
+    "/bin/rpm --quiet -e gpg-pubkey-${repo_gpg_hash_lc[${index_repo_gpg_keys}]}"
+      comment    => "This removes the GPG key from RPM if it is present and needs to be removed",
+      ifvarclass => "repo_gpg_key_${index_repo_gpg_keys}_action_remove.repo_gpg_key_${repo_gpg_hash_lc[${index_repo_gpg_keys}]}_present",
+      classes    => if_else("repo_gpg_key_${index_repo_gpg_keys}_removed", "repo_gpg_key_${index_repo_gpg_keys}_remove_error");
+
+    "/bin/rpm --quiet --import ${g.rudder_var_tmp}/gpg_key_${index_repo_gpg_keys}_content"
+      comment    => "This imports the GPG key if the necessary tmp file has been created and it needs to be imported and is not present",
+      ifvarclass => "repo_gpg_key_${index_repo_gpg_keys}_action_add.!repo_gpg_key_${repo_gpg_hash_lc[${index_repo_gpg_keys}]}_present.gpg_key_file_${index_repo_gpg_keys}_created",
+      classes    => if_else("repo_gpg_key_${index_repo_gpg_keys}_imported", "repo_gpg_key_${index_repo_gpg_keys}_import_error");
+
+  reports:
+    
+    "@@repoGpgKeyManagement@@result_repaired@@${repo_gpg_uuid[${index_repo_gpg_keys}]}@@GPG Key Management@@${repo_gpg_hash[${index_repo_gpg_keys}]}@@${g.execRun}##${g.uuid}@#The GPG Key was imported successfully" 
+      ifvarclass => "repo_gpg_key_${index_repo_gpg_keys}_imported";
```

Comment:
For the reports, you may want to use rudder_common_report (see https://github.com/Normation/rudder-techniques/blob/master/techniques/applications/aptPackageInstallation/4.0/aptPackageInstallation.st#L89 ) as it would embedeed directly support for different reporting type (change_only, full compliance..)